components/campaign/CampaignChaosBagView/index.tsx

```diff
@@ -0,0 +1,382 @@
+  }
+
+  componentDidAppear() {
+    this.setState({
+      visible: true,
+    });
+  }
+
+  componentDidDisappear() {
+    this.setState({
+      visible: false,
+    });
+  }
+
+  navigationButtonPressed({ buttonId }: { buttonId: string }) {
+    if (buttonId === 'back' || buttonId === 'androidBack') {
+      this._handleBackPress();
+    } else if (buttonId === 'edit') {
+      this._showChaosBagDialog();
+    }
+  }
+
+  _handleDrawTokenPressed = () => {
+    this._drawToken();
+  }
+
+  _handleSealTokenPressed = () => {
+
+  }
+
+  _handleBackPress = () => {
+    const {
+      componentId,
+    } = this.props;
+    const {
+      visible,
+    } = this.state;
+    if (!visible) {
+      return false;
+    }
+    Navigation.pop(componentId);
+    return true;
+  }
+
+  _showChaosBagDialog = () => {
+    const {
+      componentId,
+      updateChaosBag,
+      chaosBag,
+    } = this.props;
+
+    Navigation.push<EditChaosBagProps>(componentId, {
+      component: {
+        name: 'Dialog.EditChaosBag',
+        passProps: {
+          chaosBag,
+          updateChaosBag: updateChaosBag,
+          trackDeltas: true,
+        },
+        options: {
+          topBar: {
+            title: {
+              text: t`Chaos Bag`,
+            },
+            backButton: {
+              title: t`Cancel`,
+            },
+          },
+        },
+      },
+    });
+  }
+
+  _getRandomNumber = (min: number, max: number) => {
+    return Math.random() * (max - min) + min;
+  }
+
+  _getRandomChaosToken = (chaosBag: ChaosBag) => {
+    const list = keys(chaosBag);
+    const weight = values(chaosBag);
+
+    const totalWeight = weight.reduce((prev, cur) => {
+      return (prev && cur) ? (prev + cur) : 0;
+    });
+
+    if (totalWeight) {
+      const random_num = this._getRandomNumber(0, totalWeight);
+      let weight_sum = 0;
+      const length = list.length;
+      for (let i = 0; i < length; i++) {
+        const item = weight[i];
+        if (item) {
+          weight_sum += item;
+          weight_sum = +weight_sum.toFixed(2);
+        }
+
+        if (random_num <= weight_sum) {
+          const tokenName = list[i];
+          let drawnToken = null;
+          CHAOS_TOKENS.forEach(token => {
+            if (token === tokenName) {
+              drawnToken = token;
+            }
+          });
+
+          if (drawnToken) {
+            return drawnToken;
+          }
+        }
+      }
+    }
+  }
+
+  _drawToken = (refresh = false) => {
+    const { chaosBagResults } = this.state;
+    const { chaosBag } = this.props;
+
+    if (!chaosBagResults.currentToken || refresh) {
+      const newIconKey = this._getRandomChaosToken(chaosBag);
+      this.setState({
+        chaosBagResults: {
+          currentToken: newIconKey,
+          drawnTokens: chaosBagResults.drawnTokens,
+          total: chaosBagResults.total + 1,
+        },
+      });
+    } else {
+      this.setState({
```

Comment:
I might see if I could break this into two different functions, since they seem kind of unrelated?  Maybe I don't quite understand what's happening here -- so the if branch is if you have NO token drawn OR you are asking for a draw + refresh.  Else block is... just a refresh?

Reply:
Seems like else block is basically clearTokens, maybe just update this to call that.

Reply:
I actually refactored this already to a method named just that! `clearTokens()`